src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,805 @@
+                return items.SelectAsArray(matchResult =>
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    var highlightedSpans = _highlightMatchingPortions
+                        ? GetHighlightedSpans(matchResult, _completionHelper, _filterText)
+                        : ImmutableArray<Span>.Empty;
+
+                    return new CompletionItemWithHighlight(matchResult.EditorCompletionItem, highlightedSpans);
+                });
+
+                static ImmutableArray<Span> GetHighlightedSpans(
+                    MatchResult<VSCompletionItem> matchResult,
+                    CompletionHelper completionHelper,
+                    string filterText)
+                {
+                    if (matchResult.RoslynCompletionItem.HasDifferentFilterText)
+                    {
+                        // The PatternMatch in MatchResult is calculated based on Roslyn item's FilterText, 
+                        // which can be used to calculate highlighted span for VSCompletion item's DisplayText w/o doing the matching again.
+                        // However, if the Roslyn item's FilterText is different from its DisplayText,
+                        // we need to do the match against the display text of the VS item directly to get the highlighted spans.
+                        return completionHelper.GetHighlightedSpans(
+                            matchResult.EditorCompletionItem.DisplayText, filterText, CultureInfo.CurrentCulture).SelectAsArray(s => s.ToSpan());
+                    }
+
+                    var patternMatch = matchResult.PatternMatch;
+                    if (patternMatch.HasValue)
+                    {
+                        // Since VS item's display text is created as Prefix + DisplayText + Suffix, 
+                        // we can calculate the highlighted span by adding an offset that is the length of the Prefix.
+                        return patternMatch.Value.MatchedSpans.SelectAsArray(s_highlightSpanGetter, matchResult.RoslynCompletionItem);
+                    }
+
+                    // If there's no match for Roslyn item's filter text which is identical to its display text,
+                    // then we can safely assume there'd be no matching to VS item's display text.
+                    return ImmutableArray<Span>.Empty;
+                }
+            }
+
+            private FilteredCompletionModel? HandleAllItemsFilteredOut()
+            {
+                if (UpdateTriggerReason == CompletionTriggerReason.Insertion)
+                {
+                    // If the user was just typing, and the list went to empty *and* this is a 
+                    // language that wants to dismiss on empty, then just return a null model
+                    // to stop the completion session.
+                    if (_completionRules.DismissIfEmpty)
+                    {
+                        return null;
+                    }
```

Comment:
how is this different from the line at: https://github.com/dotnet/roslyn/pull/57685/files#diff-e43ef74201be5195fc625ca928b4c20e14e5dd304a4d42302dba2cfcc2c03557R372 ?

Reply:
This applies to all initial trigger kinds, where the dismissal in `HandleDeletionTrigger` applies to deletion trigger only (and we keep all items in the list in that case, same as handling ctrl+j). It seems we might be able to rewrite this so the dismissal in `HandleDeletionTrigger` can happen earlier (i.e. handled by this line), but I'd need to think about it more and maybe do it in a follow up PR